posteriors/utils.py

```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
     return fisher
 
 
+def _vdot_real_part(x: Tensor, y: Tensor) -> float:
+    """Vector dot-product guaranteed to have a real valued result despite
+    possibly complex input. Thus neglects the real-imaginary cross-terms.
+
+    Args:
+        x: First tensor in the dot product.
+        y: Second tensor in the dot product.
+
+    Returns:
+        The result vector dot-product, a real float
+    """
+    # all our uses of vdot() in CG are for computing an operator of the form
+    #  z^H M z
+    #  where M is positive definite and Hermitian, so the result is
+    # real valued:
+    # https://en.wikipedia.org/wiki/Definiteness_of_a_matrix#Definitions_for_complex_matrices
+    real_part = torch.einsum("...i,...i->...", x.real, y.real)
+    if torch.is_complex(x) or torch.is_complex(y):
+        imag_part = torch.einsum("...i,...i->...", x.imag, y.imag)
+        return real_part + imag_part
+    return real_part
+
+
+def _vdot_real_tree(x, y) -> TensorTree:
+    return sum(tree_leaves(tree_map(_vdot_real_part, x, y)))
+
+
+def _mul(scalar, tree) -> TensorTree:
+    return tree_map(partial(operator.mul, scalar), tree)
+
+
+_add = partial(tree_map, operator.add)
+_sub = partial(tree_map, operator.sub)
+_identity = partial(lambda x: x)
+
+
+def cg(
+    A: Callable,
+    b: TensorTree,
+    x0: TensorTree = None,
+    *,
+    maxiter: int = None,
+    damping: float = 0,
+    tol: Tensor = torch.Tensor([1e-5]),
+    atol: Tensor = torch.Tensor([0]),
+    M: Callable = _identity,
+) -> Tuple[TensorTree, Any]:
+    """Use Conjugate Gradient iteration to solve ``Ax = b``.
+    ``A`` is suplied as a function instead of a sparse
+    matrix or ``LinearOperator``.
+
+    Derivatives of ``cg`` are implemented via implicit differentiation with
+    another ``cg`` solve, rather than by differentiating *through* the solver.
+    They will be accurate only if both solves converge.
```

Comment:
Is this comment on implicit differentiation true for our implementation? Maybe just remove it